lib/Transforms/Scalar/LoopIdiomRecognize.cpp

```diff
@@ -219,6 +239,26 @@ static void deleteDeadInstruction(Instruction *I) {
 
 bool LoopIdiomRecognize::runOnLoop(Loop *L) {
   CurLoop = L;
+
+  Function *F = L->getHeader()->getParent();
+  if (sycl::isKernel(*F)) {
+    DEBUG(dbgs() << L->getName() << " in "<< F->getName()
+                 << " is kernel. Do not do LoopIdiomRecognizeLegacyPass.\n");
+    return false;
+  }
+
+  // If function is used by kernel, skip this pass.
+  for (Use &U : F->uses()) {
+    CallSite CS(U.getUser());
+    if (CS.getInstruction() != nullptr) {
+      if (sycl::isKernel(*(CS.getInstruction()->getParent()->getParent()))) {
+        DEBUG(dbgs() << L->getName() << " in function that is used by kernel. "
+                     << "Do not do LoopIdiomRecognizeLegacyPass.\n");
+        return false;
+      }
+    }
+  }
+
```

Comment:
No, code added in this file is to force generated IR to be what argument promotion pass liked. Since argument promotion can only be done with GEP instructions that
only used by loads. And if there are llvm.memcpy instruction, there
will always a bitcast instruction before to cast the pointer. This makes argument promotion not working, so force kernels and functions called in the kernel in per_kernel_caller.ll to
skip loop idiom opt pass.